sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala

```diff
@@ -74,3 +86,321 @@ private[hive] trait HiveStrategies {
+    val defaultSource = new OrcFileFormat()
+    val fileFormatClass = classOf[OrcFileFormat]
+    val options = Map[String, String]()
+
+    ConvertMetastoreTablesUtils.convertToLogicalRelation(
+      sparkSession, relation, options, defaultSource, fileFormatClass, "orc")
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!plan.resolved || plan.analyzed) {
+      return plan
+    }
+
+    plan transformUp {
+      // Write path
+      case InsertIntoTable(r: MetastoreRelation, partition, child, overwrite, ifNotExists)
+        // Inserting into partitioned table is not supported in Orc data source (yet).
+        if !r.hiveQlTable.isPartitioned && shouldConvertMetastoreOrc(r) =>
+        InsertIntoTable(convertToOrcRelation(r), partition, child, overwrite, ifNotExists)
+
+      // Read path
+      case relation: MetastoreRelation if shouldConvertMetastoreOrc(relation) =>
+        val orcRelation = convertToOrcRelation(relation)
+        SubqueryAlias(relation.tableName, orcRelation)
+    }
+  }
+}
+
+/**
+ * When scanning or writing to non-partitioned Metastore Parquet tables, convert them to Parquet
+ * data source relations for better performance.
+ */
+class ParquetConversions(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+  private def shouldConvertMetastoreParquet(relation: MetastoreRelation): Boolean = {
+    relation.tableDesc.getSerdeClassName.toLowerCase.contains("parquet") &&
+      sparkSession.sessionState.conf.getConf(HiveUtils.CONVERT_METASTORE_PARQUET)
+  }
+
+  private def convertToParquetRelation(relation: MetastoreRelation): LogicalRelation = {
+    val defaultSource = new ParquetFileFormat()
+    val fileFormatClass = classOf[ParquetFileFormat]
+
+    val mergeSchema = sparkSession.sessionState.conf.getConf(
+      HiveUtils.CONVERT_METASTORE_PARQUET_WITH_SCHEMA_MERGING)
+    val options = Map(ParquetOptions.MERGE_SCHEMA -> mergeSchema.toString)
+
+    ConvertMetastoreTablesUtils.convertToLogicalRelation(
+      sparkSession, relation, options, defaultSource, fileFormatClass, "parquet")
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!plan.resolved || plan.analyzed) {
+      return plan
+    }
+
+    plan transformUp {
+      // Write path
+      case InsertIntoTable(r: MetastoreRelation, partition, child, overwrite, ifNotExists)
+        // Inserting into partitioned table is not supported in Parquet data source (yet).
+        if !r.hiveQlTable.isPartitioned && shouldConvertMetastoreParquet(r) =>
+        InsertIntoTable(convertToParquetRelation(r), partition, child, overwrite, ifNotExists)
+
+      // Read path
+      case relation: MetastoreRelation if shouldConvertMetastoreParquet(relation) =>
+        val parquetRelation = convertToParquetRelation(relation)
+        SubqueryAlias(relation.tableName, parquetRelation)
+    }
+  }
+}
+
+object ConvertMetastoreTablesUtils extends Logging {
```

Comment:
If we can combine the rules `OrcConversions` and `ParquetConversions` into the same rule, this object can be removed.